test/uvw/tcp.cpp

```diff
@@ -148,3 +148,56 @@ TEST(TCP, WriteError) {
     ASSERT_LT(handle->try_write(std::unique_ptr<char[]>{}, 0), 0);
     ASSERT_LT(handle->try_write(nullptr, 0), 0);
 }
+
+TEST(TCP, CustomAllocator) {
+    const std::string address = std::string{"127.0.0.1"};
+    const unsigned int port = 4242;
+
+    auto loop = uvw::loop::get_default();
+    auto server = loop->resource<uvw::tcp_handle>();
+    auto client = loop->resource<uvw::tcp_handle>();
+
+    auto *ptr = new char[5];
+
+    server->on<uvw::error_event>([](const auto &, auto &) { FAIL(); });
+    client->on<uvw::error_event>([](const auto &, auto &) { FAIL(); });
+
+    server->on<uvw::listen_event>([ptr](const uvw::listen_event &, uvw::tcp_handle &handle) {
+        std::shared_ptr<uvw::tcp_handle> socket = handle.parent().resource<uvw::tcp_handle>();
+
+        socket->on<uvw::error_event>([](const uvw::error_event &, uvw::tcp_handle &) { FAIL(); });
+        socket->on<uvw::close_event>([&handle](const uvw::close_event &, uvw::tcp_handle &) { handle.close(); });
+        socket->on<uvw::end_event>([](const uvw::end_event &, uvw::tcp_handle &sock) { sock.close(); });
+        socket->on<uvw::data_event>([ptr](const uvw::data_event &evt, uvw::tcp_handle &sock) {
+            auto &d = const_cast<uvw::data_event &>(evt);
+            auto *data = d.data.release();
+            ASSERT_LE(evt.length, 5);
+            ASSERT_EQ(data, ptr);
+            sock.close();
+        });
+        socket->allocator = [ptr](uv_buf_t *b, size_t suggested, uvw::tcp_handle &h) {
```

Comment:
The idea was a little different. For example, since `uv_read_start` also accepts an allocator, `stream_handle<...>::read` can accept an allocator function as well (with a default) and forward it to the underlying function. Why the proposed solution should be better instead?

Reply:
In my solution user can change allocator in event handlers. Also it allows to setup allocator once and do `start()`-`stop()` loops without passing custom function every time.

Reply:
Also forwarding directly to `uv_read_start` limits allocator implementation to real functions only. No lambda with capturing list

Reply:
Unfortunately, this is just an illusion if not even a mistake.

Imagine to set an allocator, then start a read request and finally change the allocator immediately after.
Because of how `libuv` works, the last allocator is invoked, although reading the code makes you think that the first one is used instead.
Now imagine to change the allocator _somewhere else_ in the codebase rather than _immediately after_. Maybe a couple of months after your change. Possibly a commit from a colleague or a PR from a random guy on the globe who doesn't know well how things work and what the expectation is.
I can already hear the curses of the dev who's called to fix this subtle bug. 😅 

When it comes to functional programming, passing something to the function that triggers the call is 99% of the time right or, well, at least predicatable, while _storing aside something for later uses_ gives a false feeling on what's going to happen and is more often than not a recipe for disasters.

Reply:
> although reading the code makes you think that the first one is used instead.

Well, for me it doesn't make.

I understand your points, but for me it is important to store alllocator. Will wait for your implementation.